aten/src/ATen/native/transformers/attention.cpp

```diff
@@ -379,102 +441,45 @@ std::tuple<Tensor, Tensor> native_multi_head_attention(
+  c10::optional<Tensor> attn_mask;
+  if (mask.has_value()) {
+      // take mask from (B, T) -> (B, 1, 1, T) to match 4D inputs in (B, num_head, T, T) format
+      // Note: SDP has opposite masking semantics
+      attn_mask = (mask->dim() == 2 ? mask->unsqueeze(1).unsqueeze(1) : *mask).logical_not();
   }
-#endif
-  // shape: 3 x [B, num_head, T, dim_per_head]
-  auto q_k_v = _transform_bias_rescale_qkv(qkv, qkv_bias, num_head);
-  qkv = Tensor(); // Not used any more, allow free
-  auto& q = std::get<0>(q_k_v);
-  const auto& k = std::get<1>(q_k_v);
-  const auto& v = std::get<2>(q_k_v);
-#ifndef NDEBUG
-  debug_assert_shape(__LINE__, q, {B, num_head, T, dim_per_head});
-  debug_assert_shape(__LINE__, k, {B, num_head, T, dim_per_head});
-  debug_assert_shape(__LINE__, v, {B, num_head, T, dim_per_head});
-#endif
-#ifdef DEBUG_PRINT_EACH_STEP
-  std::cerr << "q: " << q << std::endl;
-  std::cerr << "k: " << k << std::endl;
-  std::cerr << "v: " << v << std::endl;
-#endif
+  std::tie(attn_output, attn_weights) = at::_scaled_dot_product_attention(
+          q, k, v, attn_mask, /*dropout=*/ 0.0, /*need_attn_weights=*/ need_weights);
 
-  // shape: [B, num_head, T, T]
-  auto qkt = bmm_nt(q, k);
-  // q & k are dead but cannot be freed because they were packed with v
-#ifndef NDEBUG
-  debug_assert_shape(__LINE__, qkt, {B, num_head, T, T});
-#endif
-#ifdef DEBUG_PRINT_EACH_STEP
-  std::cerr << "qkt: " << qkt << std::endl;
-#endif
+  // Convert attn_output from shape [B, num_head, T, dim_per_head] -> [B, T, D]
+  attn_output = attn_output.transpose(1, 2).contiguous().reshape({B, -1, D});
+  auto proj = at::linear(attn_output, proj_weight, proj_bias);
 
-  // shape: [B, num_head, T, T]
-  // TODO: long-term, have a kernel that works with
-  // NestedTensor directly if there is no mask passed
-  qkt = masked_softmax(qkt, mask, query, mask_type);
-#ifdef DEBUG_PRINT_EACH_STEP
-  std::cerr << "qkt after softmax: " << qkt << std::endl;
-#endif
-
-  // shape: [B, num_head, T, dim_per_head]
-  // reuse storage for q; we're done with it
-  auto attn_ctx = bmm_nn(q, qkt, v);
-  // qkv is not dead; we just reused storage for q!
-  if (!need_weights) {
-    qkt = Tensor();
-  }
-#ifndef NDEBUG
-  debug_assert_shape(__LINE__, attn_ctx, {B, num_head, T, dim_per_head});
-#endif
-#ifdef DEBUG_PRINT_EACH_STEP
-  std::cerr << "attn_ctx: " << attn_ctx << std::endl;
-#endif
-
-  // shape: [B, T, D]
-  // Fuse transform_0213 inside
-  auto proj = transform0213_gemm_nt_bias(
-      attn_ctx, proj_weight, proj_bias, query);
-#ifndef NDEBUG
-  debug_assert_shape(__LINE__, proj, {B, T, D});
-#endif
   if (need_weights && average_attn_weights) {
     // weights are not needed for full transformer, so don't worry too
     // much about performance -- we implement this just to make use
     // cases that don't disable need_weights still get some speedup.
-    qkt = qkt.sum(1);
-    qkt /= num_head;
+    attn_weights = attn_weights.sum(1);
```

Comment:
This is causing an error right now because attn_weights is a nested tensor and can only be reduced across last dim